sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -335,12 +343,30 @@ abstract class TypeCoercionBase {
       // Return the result after the widen data types have been found for all the children
       if (attrIndex >= children.head.output.length) return castedTypes.toSeq
 
-      // For the attrIndex-th attribute, find the widest type
-      val widenTypeOpt = findWiderCommonType(children.map(_.output(attrIndex).dataType))
```

Comment:
I think it's clearer if `CollationTypeCasts` is the only rule doing string collation type coercion. Should `findWiderCommonType` return None if there are string collations, so that we can handle it later in `CollationTypeCasts`?

Reply:
These changes would require us to override `CollationTypeCasts.apply` method and there add a special case for these set operators with collated input types. If that is what you would also recommend, I will proceed with that approach and refactor the `CollationTypeCast`'s apply method